### ISSUE TITLE
Caching the results of running prepare_command can cause issues.

### ISSUE THREAD

**Issue by iosdevzone** [CocoaPods/CocoaPods#7192]
# Report

CocoaPods caches any files produced by the `prepare_command`. If any of these files depend on the version of Xcode a user is running this can cause stale files to be installed as part of a pod. Removing the `Podfile.lock` file and `Pods` directory is not sufficient (though many users believe it is). The issue is somewhat exacerbated by the fact that without the `--verbose` flag CocoaPods does not mention that it is using a cached copy of the pod. Indeed, consider the following output:

````
Analyzing dependencies
Downloading dependencies
Installing IDZSwiftCommonCrypto (0.9.1)
Generating Pods project
Integrating client project
Sending stats
Pod installation complete! There is 1 dependency from the Podfile and 1 total pod installed.
````
seems reasonable to conclude that the pod had been downloaded.

Following the principle of least surprise it seems to me that it would be better if CocoaPods cached the state of the pod *before* the `prepare_command` had been run and ran it on the cached copy after installation in the user's directory.

I know this is probably only an issue for a small number of pods, but it took quite some time to diagnose the issue. To get an idea of the sort of confusion this can cause see: IDZSwiftCommonCrypto Issues: [51](https://github.com/iosdevzone/IDZSwiftCommonCrypto/issues/51) [78](https://github.com/iosdevzone/IDZSwiftCommonCrypto/issues/78).

## What did you do?

1) Upgraded Xcode
2) Removed `Podfile.lock` and `Pods` directory
3) Ran `pod install`

## What did you expect to happen?

The pod installation should be identical to what I would have obtained had the pod not been in the cache.

## What happened instead?

Since the cached version of the pod contained results of running `prepare_command` with an older version of Xcode come files in the pod were out of date.

### Stack

```
   CocoaPods : 1.2.1
        Ruby : ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-darwin15]
    RubyGems : 2.6.8
        Host : Mac OS X 10.13.1 (17B35a)
       Xcode : 9.1 (9B55)
         Git : git version 2.13.6 (Apple Git-96)
Ruby lib dir : /Users/danny/.rvm/rubies/ruby-2.3.3/lib
Repositories : fft-specs - /Users/danny/fft/apps/Specs.git @ 0d92977b59385238cf908d2819e03b6f433eeed2
               IDZPodspecs - https://github.com/iosdevzone/IDZPodspecs.git @ 7c0aefd6723b21eec2efe0ff6c2c5e82a9db1743
               iosdevzone - https://github.com/iOSDevZone/IDZPodspecs.git @ 7c0aefd6723b21eec2efe0ff6c2c5e82a9db1743
               master - https://github.com/CocoaPods/Specs.git @ 83159e39295ad4f576c4524f10e57e6de4fc1aed
```

### Installation Source

```
Executable Path: /Users/danny/.rvm/gems/ruby-2.3.3/bin/pod
```

### Plugins

```
claide-plugins        : 0.9.2
cocoapods-deintegrate : 1.0.1
cocoapods-playgrounds : 1.2.2
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-stats       : 1.0.0
cocoapods-trunk       : 1.2.0
cocoapods-try         : 1.1.0
```

## Project that demonstrates the issue

The issue is demonstrated by:
https://github.com/iosdevzone/IDZSwiftCommonCrypto
even using `xcode-select` to switch active Xcodes can demonstrate the problem. See the generated `module.map` files.

**Comment by dnkoutso** [comment]
Prepare commands are only supposed to be executed once during fresh installation of the pod. I see your concern and problem though.

Would it be better to use a `script_phase` instead to execute this as part of Xcode compilation since it depends on Xcode version? Script phase support is being added in 1.4.0.

You can use 1.4.0.beta.2 which has support but might need to ensure your podspec specifies `s.cocoapods_version '>= 1.4.0.beta.2'`

**Comment by dnkoutso** [comment]
@iosdevzone see https://github.com/CocoaPods/Core/blob/master/lib/cocoapods-core/specification/dsl.rb#L1285-L1322

**Comment by iosdevzone** [comment]
Yup, sounds like `script_phase` is exactly what I need. Thanks for the info! The Xcode project uses a script phase so it's a closer match to what I want. Don't want to push users to rely on beta CocoaPods so will wait until it is in a released version.

Still think the caching behavior is a bit surprising. Perhaps a mention in documentation for `prepare_command` highlighting "download" vs. "install from cache" would be useful.

**Comment by dnkoutso** [comment]
@iosdevzone I marked it as an easy enhancement. No plans on implementing this soon. Anyone who feels contributing can jump right in!
